chebi-update/src/main/java/org/reactome/release/chebiupdate/ChebiDataRetriever.java

```diff
@@ -0,0 +1,267 @@
+		else
+		{
+			// Other Webservice errors should probably break execution - if one fails, they will all probably fail.
+			// This is *not* a general principle, but is based on my experience with the ChEBI webservice specifically -
+			// it's a pretty stable service so it's unlikely that if one service call fails, the others will succeed.
+			logger.error("WebService error occurred! Message is: {}", e.getMessage());
+			e.printStackTrace();
+			throw new RuntimeException(e);
+		}
+	}
+
+	/**
+	 * Get data from the ChEBI web service.
+	 * @param failedEntitiesList - a list of failed entities, this may be updated if the query to ChEBI fails.
+	 * @param chebiCache - the cache of ChEBI identifiers, this will be updated if the query is successful.
+	 * @param bw - writer for the cache.
+	 * @param molecule - the Molecule to query for.
+	 * @return - The Entity, from ChEBI.
+	 * @throws ChebiWebServiceFault_Exception
+	 * @throws IOException
+	 * @throws Exception
+	 */
+	private Entity getChEBIDataFromWebService(Map<GKInstance, String> failedEntitiesList, final Map<String, List<String>> chebiCache, BufferedWriter bw, GKInstance molecule) throws ChebiWebServiceFault_Exception, IOException, Exception
+	{
+		String identifier = (String) molecule.getAttributeValue(ReactomeJavaConstants.identifier);
+		// If there's a valid cache and we're supposed to use it, log a message here indicating
+		// that there must have been a cache miss.
+		if (this.useCache && chebiCache.size() > 0)
+		{
+			logger.trace("Cache miss for CHEBI:{}", identifier);
+		}
+		Entity entity = this.chebiClient.getCompleteEntity(identifier);
+		// IF there is valid entity AND we are supposed to use a cache, then write the entity to the cache file.
+		if (entity != null && this.useCache)
+		{
+			bw.write("CHEBI:"+identifier+"\t"+entity.getChebiId()+"\t"+entity.getChebiAsciiName()+"\t"+ (entity.getFormulae().size() > 0 ? entity.getFormulae().get(0).getData() : "") + "\t" + LocalDateTime.now().toString() + "\n");
+			bw.flush();
+		}
+		// IF the entity was null, then log a message that the webservice response was null!
+		if (entity == null)
+		{
+			failedEntitiesList.put(molecule, "ChEBI WebService response was NULL.");
+		}
+		
+		return entity;
+	}
+
+	/**
+	 * Returns the data from the cache file (if it exists). If it doesn't exit, then an empty map will be returned.
+	 * @return A mapping of ChEBI ID to a list with items in this order: ChEBI (might be different if ChEBI has changed identifiers), ChEBI Name, Formula.
+	 * @throws IOException
+	 */
+	private Map<String,List<String>> loadCacheFromFile() throws IOException
+	{
+		Map<String,List<String>> chebiCache = Collections.synchronizedMap(new HashMap<String, List<String>>());
+		if (this.useCache)
+		{
+			logger.info("useCache is TRUE - chebi-cache file will be read, and populated. Identifiers not in the cache will be queried from ChEBI.");
+			// if the cache exists, load it.
+			if (Files.exists(Paths.get("chebi-cache")))
```

Comment:
Should 'chebi-cache' string be stored in a variable since it is used multiple times?

Reply:
:thinking: